File: HW1/task1.scala
import org.apache.spark.{SparkContext,SparkConf}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import java.io._
// import scala.io.Source

object task1 {
  def main(arg:Array[String]): Unit={
    /*
    Initializing
    */
    val input_file = arg(0)
    val output_file = arg(1)
    val spark = new SparkConf().setAppName("task1").setMaster("local[*]")
    val sc = new SparkContext(spark)
    /*
    Reading file
    */

    val RDD = sc.textFile(input_file).map{ row =>
          val json_row = parse(row)
          (json_row)
     }.cache()

    /*
    Task1
    */
    // val res = Map[String, Any]()
    // A. total number of reviews
    // res += ("n_review" -> RDD.count())
    // B. number of reviews in 2018
    // res += ("n_review_2018" -> RDD.filter(x => pretty(render(x \"date")).slice(1, 5) == "2018").count())
    // C. number of distinct users
    // res += ("n_user" -> RDD.map(x => (x \"user_id")).distinct.count())
    // D. top ten users with most reviews
    // res += ("top10_user" -> RDD.map(x => (pretty(render(x \"user_id")), 1)).reduceByKey(_+_).sortBy(x => (-x._2, x._1)).collect.toList.slice(0, 10))
    // E. number of distinct businesses
    // res += ("n_business" -> RDD.map(x => pretty(render(x \"business_id"))).distinct.count())
    // F. top ten businesses with most reviews
    // res += ("top10_business" -> RDD.map(x => (pretty(render(x \"business_id")), 1)).reduceByKey(_+_).sortBy(x => (-x._2, x._1)).collect.toList.slice(0, 10))
    // val counts = text.flatMap(line => line.split(" ")).map(word => (word,1)).reduceByKey(_+_)
    val res = Map(
      "n_review" -> RDD.count(),
      "n_review_2018" -> RDD.filter(x => pretty(render(x \"date")).slice(1, 5) == "2018").count(),
      "n_user" -> RDD.map(x => (x \"user_id")).distinct.count(),
      "top10_user" -> RDD.map(x => (pretty(render(x \"user_id")), 1)).reduceByKey(_+_).sortBy(x => (-x._2, x._1)).collect.toList.slice(0, 10),
      "n_business" -> RDD.map(x => pretty(render(x \"business_id"))).distinct.count(),
      "top10_business" -> RDD.map(x => (pretty(render(x \"business_id")), 1)).reduceByKey(_+_).sortBy(x => (-x._2, x._1)).collect.toList.slice(0,10)
    )
    // wrtie file
    val file = new File(output_file)
    val bw = new BufferedWriter(new FileWriter(file))
    bw.write(scala.util.parsing.json.JSONObject(res).toString().replace("List(","[").replace("(","[").replace(")","]"))
    bw.close()
  }
}

File: HW1/task1.py
# HW1 task1
# Po-Han Chen USC ID:

import json
from pyspark import SparkContext
import os
import sys

# parse command line arguments
input_path = sys.argv[1]
output_path = sys.argv[2]

# configuration on local machine
os.environ['PYSPARK_PYTHON'] = '/usr/local/bin/python3.6'
os.environ['PYSPARK_DRIVER_PYTHON'] = '/usr/local/bin/python3.6'


def load_json(string):
    """
    Drop text to save space and time
    """
    json_data = json.loads(string)
    return json_data


# read part of the file
# with open(input_path) as in_file:
    # data = in_file.readlines(2000000)
sc = SparkContext('local[*]', 'task1')
# RDD = sc.parallelize(data).map(load_json).cache()
RDD = sc.textFile(input_path).map(load_json).cache()
"""
Task 1 
"""
res = dict()
# A. count
# print("-"*10+"count of reviews: "+str(RDD.count()))
res['n_review'] = RDD.count()
# B. number of reviews in 2018
# print(RDD.filter(lambda x: x['date'][0:4] == '2018').count())
res['n_review_2018'] = RDD.filter(lambda x: x['date'][0:4] == '2018').count()
# C. number of distinct users who wrote reviews
# print("-"*10+'distinct users: '+str(RDD.map(lambda x: x['user_id']).distinct().count()))
res['n_user'] = RDD.map(lambda x: x['user_id']).distinct().count()
# D. top 10 users with most reviews
# print("-"*10+'top 10 users: '+str(RDD.map(lambda x: (x['user_id'], 1)).reduceByKey(lambda x, y: x+y).
# sortBy(lambda x: x[1], ascending=False).take(10)))
res['top10_user'] = RDD.map(lambda x: (x['user_id'], 1)).reduceByKey(lambda x, y: x+y).\
                                  sortBy(lambda x: (-x[1], x[0])).take(10)
# E. number of distinct businesses
# print("-"*10+'distinct businesses: '+str(RDD.map(lambda x: x['business_id']).distinct().count()))
res['n_business'] = RDD.map(lambda x: x['business_id']).distinct().count()
# F. top ten businesses and their reviews
# print("-"*10+'top 10 businesses: '+str(RDD.map(lambda x: (x['business_id'], 1)).reduceByKey(lambda x, y: x+y).
# sortBy(lambda x: x[1], ascending=False).take(10)))
res['top10_business'] = RDD.map(lambda x: (x['business_id'], 1)).reduceByKey(lambda x, y: x+y).\
                                sortBy(lambda x: (-x[1], x[0])).take(10)
# output file
with open(output_path, 'w') as out_file:
    json.dump(res, out_file)

